### ISSUE TITLE
Exploring Automated Dispatching of DataLoader in Ballerina

### ISSUE THREAD

**Issue by MohamedSabthar** [ballerina-platform/ballerina-standard-library#4302]
### The problem with GraphQL and N+1 queries

The N+1 problem is a performance issue that can occur when querying data from a GraphQL API. It arises when a single GraphQL query triggers multiple database queries, resulting in significant overhead and reduced performance. The problem is due to the flexible data model of GraphQL, which allows clients to specify the exact fields they want to retrieve in a query. To avoid the N+1 problem, the DataLoader library provides mechanisms like batching and caching which allows multiple database queries to be executed in a single round trip.

##### Example javascript graphql code to demonstrate N+1 queries

```javascript
import { ApolloServer } from "@apollo/server";
import { startStandaloneServer } from "@apollo/server/standalone";
import query from "./query.js";
import mysql from "mysql2";

const typeDefs = `#graphql
  type Query {
    authors: [Author!]!
  }
  
  type Book {
    id: Int!
    title: String!
  }
  
  type Author {
    name: String!
    books: [Book!]!
  }
`;

const resolvers = {
  Query: {
    authors: () => {
      const sql = `SELECT * FROM authors`;
      console.log(mysql.format(sql));
      return query(sql);
    },
  },
  Author: {
    books: (author) => {
      const sql = `SELECT * FROM books WHERE author = ?`;
      const params = [author.id];
      console.log(mysql.format(sql, params));
      return query(sql, params);
    },
  },
};

const server = new ApolloServer({
  typeDefs,
  resolvers,
});

const { url } = await startStandaloneServer(server, {
  listen: { port: 9090 },
});

console.log(`🚀  Server ready at: ${url}`);
```
The full code for this example can be found [here](https://github.com/MohamedSabthar/gql-loader-samples/tree/n%2B1/simple-js)

Querying the following graphql query operation

```graphql
query Query {
  authors {
    books {
      title
    }
  }
}
```

for the above javascript service will result in following following N+1 db queries
```sql
SELECT * FROM authors              # returns 10 authors
SELECT * FROM books WHERE author = 1
SELECT * FROM books WHERE author = 2
SELECT * FROM books WHERE author = 3
SELECT * FROM books WHERE author = 4
SELECT * FROM books WHERE author = 5
SELECT * FROM books WHERE author = 6
SELECT * FROM books WHERE author = 7
SELECT * FROM books WHERE author = 8
SELECT * FROM books WHERE author = 9
SELECT * FROM books WHERE author = 10 
# a separate query is executed foreach author therefore N+1 queries
```

### How does the DataLoader library work?

The javascript DataLoader library works by allowing a GraphQL resolver function to defer a database query and return a `promise` object to the GraphQL server. The DataLoader library then collects all deferred queries from the resolver functions and batches them into a single round trip to the database.

When the batched queries are executed, DataLoader caches the results and returns the resolved promise objects to the original resolver functions. This ensures that each query is executed only once, and subsequent requests for the same data are served from the cache.

#### Example javascript graphql code to demonstrate batching with DataLoader

```javascript
// This is a user defined batch function provided to the DataLoader constructor
async function batchBooks(authorIds) {
  const sql = `SELECT * FROM books WHERE author IN (?)`;
  const params = [authorIds];
  const books = await query(sql, params);
  const grouped = ramda.groupBy((book) => book.author, books);
  return ramda.map((id) => grouped[id] || [], authorIds);
}

const resolvers = {
  Query: {
    authors: () => getAuthors(),
  },
  Author: {
    books: ({ id }, _, { bookLoader }) => bookLoader.load(id).then((books)=> books),
    // Following commented code is an alternative to the above line.
    // books: async ({ id }, _, { bookLoader }) => {
    //   var books = await bookLoader.load(id);
    //   return books;
    // },
  },
};

const { url } = await startStandaloneServer(server, {
  listen: { port: 9090 },

  context: () => {
    return {
      bookLoader: new DataLoader((keys) => batchBooks(keys))
    };
  },
})

console.log(`🚀  Server ready at: ${url}`);
```
The full code for this example can be found [here](https://github.com/MohamedSabthar/gql-loader-samples/tree/n%2B1/loader-js)

##### Load method

The load method of the DataLoader takes a single key as input and returns a Promise that resolves to the corresponding value for that key. If the value for the key is not already in the cache, DataLoader will queue a new batched query for that key, which will be dispatched at the end of the current event loop.

##### BatchLoad function

In the example, the batchBooks function is a function defined by the user, which serves as the batchLoad function that is supplied as the first argument to the DataLoader constructor. Its purpose is to batch and load data for multiple keys simultaneously. Whenever a load method is invoked with a particular key, the DataLoader instance queues up that key for deferred database query execution. Then, all of the queued keys are sent to the batchLoad function upon DataLoader batch dispatch.

querying the same graphql query operation as above will result in following db queries
```sql
SELECT * FROM authors   # returns 10 authors
SELECT * FROM books WHERE author IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
# all books are fetched in a single query using the DataLoader batching mechanism
```

### When does the DataLoader dispatch a batched (database) query?

DataLoader dispatches a batch of deferred queries to the underlying data source when the current execution context is about to be completed. This ensures that DataLoader has collected all deferred queries from resolver functions and can batch them into a single round trip (using the user written batch function) to the database. The batch dispatching mechanism is triggered automatically by DataLoader and dispatch logic is hidden from the resolvers.

Is there a way to achieve this functionality in ballerina ?

**Comment by MohamedSabthar** [comment]
Move to ballerina-spec repo
